### ISSUE TITLE
Improve `EventLog`

### ISSUE THREAD

**Issue by alexandrepiveteau** [markdown-party/kotlin-echo#16]
There are multiple flaws with the current `EventLog` interfaces :

- It's not possible to access all the events of a single site easily.
- `toPersistentLog` is not an extension function.
- `fold` is not an extension function and must be defined on the data type.
- `expected` has a strange semantic.

This could be solved as follows :

- Introducing an abstraction layer for sites (`ImmutableSiteEventLog` or `ImmutableSiteLog`).
- Making `toPersistentLog` an extension function.
- Defining a generic way to combine all sites from a log, and then their events in the right order.
- Having a `next()` function on both `ImmutableEventLog` and `ImmutableSiteEventLog`.

---
# Code exploration :

```kotlin
package markdown.echo.logs2

import markdown.echo.causal.EventIdentifier
import markdown.echo.causal.SequenceNumber
import markdown.echo.causal.SiteIdentifier

interface ImmutableEventLog<out T> {

  val sites: Set<SiteIdentifier>

  fun next(site: SiteIdentifier): EventIdentifier

  operator fun get(site: SiteIdentifier): ImmutableSiteLog<T>
  operator fun get(site: SiteIdentifier, seqno: SequenceNumber): T?
}

interface PersistentEventLog<out T> : ImmutableEventLog<T> {

  fun set(
      site: SiteIdentifier,
      seqno: SequenceNumber,
      body: @UnsafeVariance T,
  ): PersistentEventLog<T>
}

fun <T> ImmutableEventLog<T>.toPersistentEventLog(): PersistentEventLog<T> {
  if (this is PersistentEventLog) return this else TODO("Add from all sites.")
}

interface ImmutableSiteLog<out T> : Collection<T> {

  val site: SiteIdentifier

  fun next(): SequenceNumber

  operator fun get(seqno: SequenceNumber): T?
}

// Folding the sequence takes O(n * k) time. Each step takes O(k) amortized to compute.
fun <T : Comparable<T>, R> ImmutableEventLog<T>.fold(initial: R, f: (R, T) -> R): R {
  val mapped = sites.asSequence().map { get(it).asSequence() }.toList().toTypedArray()
  return combine(*mapped).fold(initial, f)
}

fun <C : Comparable<C>> combine(a: Sequence<C>, b: Sequence<C>): Sequence<C> = sequence {
  val aIterator = a.iterator()
  var aConsumed = true
  var aHasNext = aIterator.hasNext()
  lateinit var aElem: C

  val bIterator = b.iterator()
  var bConsumed = true
  var bHasNext = bIterator.hasNext()
  lateinit var bElem: C

  while (aHasNext && bHasNext) {
    if (aConsumed) {
      aElem = aIterator.next()
      aHasNext = aIterator.hasNext()
      aConsumed = false
    }

    if (bConsumed) {
      bElem = bIterator.next()
      bHasNext = bIterator.hasNext()
      bConsumed = false
    }

    // We have the latest value for both sites.
    if (aElem < bElem) {
      aConsumed = true
      yield(aElem)
    } else {
      bConsumed = true
      yield(bElem)
    }
  }

  // TODO : Ajouter les éléments restants d'une ou de l'autre séquence.

}

fun <C : Comparable<C>> combine(vararg elements: Sequence<C>): Sequence<C> {

  tailrec fun helper(
      acc: Sequence<C>,
      next: Int,
      vararg elements: Sequence<C>,
  ): Sequence<C> {
    if (next == elements.size) return acc
    return helper(combine(acc, elements[next]), next + 1, *elements)
  }

  return helper(emptySequence(), 0, *elements)
}

fun main() {
  val a: Iterable<Int> = listOf(1, 2, 3)
  val b = sequenceOf(4, 5, 6)
}
```

**Comment by alexandrepiveteau** [comment]
Being worked on in #33.

**Comment by alexandrepiveteau** [comment]
Fixed in #83.
